cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -1130,151 +1217,198 @@ aggregate_reader_metadata::select_row_groups(
+          auto const num_selected_rows_this_row_group =
+            is_trimmed_row_groups ? row_group_num_rows[src_idx][rg_idx] : rg.num_rows;
+
+          // Update the number of rows read from this data source and the total number of selected
+          // rows
+          num_rows_per_source[src_idx] += num_selected_rows_this_row_group;
+          total_selected_rows += num_selected_rows_this_row_group;
 
           // We need the unadjusted start index of this row group to correctly initialize
-          // ColumnChunkDesc for this row group in create_global_chunk_info() and calculate
-          // the row offset for the first pass in compute_input_passes().
-          selection.emplace_back(rg_idx, chunk_start_row, src_idx);
+          // ColumnChunkDesc for this row group in create_global_chunk_info() and calculate the
+          // row offset for the first pass in compute_input_passes().
+          selection.emplace_back(rg_idx, row_group_start_row, src_idx);
 
           // If page-level indexes are present, then collect extra chunk and page info.
-          // The page indexes rely on absolute row numbers, not adjusted for skip_rows.
-          column_info_for_row_group(selection.back(), chunk_start_row);
-        }
-        // Adjust the number of rows for the last source file.
-        if (count >= rows_to_skip + rows_to_read) {
-          num_rows_per_source[src_idx] -= count - rows_to_skip - rows_to_read;
-        }
-      }
-    }
+          // The page indexes rely on absolute row numbers - not adjusted for skip_rows - or use
+          // zero as start row if row groups were filtered using stats and bloom filters.
+          column_info_for_row_group(selection.back(),
+                                    is_filtered_row_groups ? 0 : row_group_start_row);
 
-    // If filter had a value and no row groups were filtered, set the number of row groups after
-    // filters to the number of adjusted input row groups
-    auto const after_stats_filter = num_row_groups_after_filters.after_stats_filter.has_value()
-                                      ? std::make_optional(total_row_groups)
-                                      : std::nullopt;
-    auto const after_bloom_filter = num_row_groups_after_filters.after_bloom_filter.has_value()
-                                      ? std::make_optional(total_row_groups)
-                                      : std::nullopt;
-    num_row_groups_after_filters  = {after_stats_filter, after_bloom_filter};
-  }
+          // Disable the is_first_row_group flag
+          is_first_row_group = false;
+                                });
+                });
+
+  // Set the final number of rows to read to the number of selected rows
+  rows_to_read = total_selected_rows;
 
   return {rows_to_skip,
           rows_to_read,
           std::move(selection),
           std::move(num_rows_per_source),
-          total_row_groups,
+          total_row_bounds_filtered_row_groups,
```

Comment:
Note that this is the same `total_row_groups` as before which was the total row groups after trimming row groups with row bounds (skip_rows/num_rows).